File: chess/moveset.go
package chess

// MoveSet set of possible moves for a piece
type MoveSet struct {
	moves map[uint]Move
}

// MakeMoveSet create a new moveset
func MakeMoveSet() MoveSet {
	return MoveSet{ make(map[uint]Move) }
}

// Clone deep clone the move set
func (set MoveSet) Clone() MoveSet {
	clone := MoveSet{ make(map[uint]Move) }
	for id, move := range set.moves {
		clone.moves[id] = move
	}
	return clone
}

// Add add the move to the moveset
func (set MoveSet) Add(move Move) {
	set.moves[move.ID()] = move
}
func (set MoveSet) addID(id uint, move Move) {
	set.moves[id] = move
}

// IsEmpty return true if the moveset is empty
func (set MoveSet) IsEmpty() bool {
	return len(set.moves) == 0
}

// GetMove return a move based on the position (false if no move found)
func (set MoveSet) GetMove(pos Vec2) (Move, bool) {
	return set.getID(pos.ID())
}
func (set MoveSet) getID(id uint) (Move, bool) {
	move, ok := set.moves[id]
	return move, ok
}

// Purge remove all invalid moves from the moveset
func (set MoveSet) Purge(board Board, piece Piece) {
	for id, move := range set.moves {
		tempBoard := board.Clone()
		tempPiece := piece
		move.Apply(tempBoard, tempPiece)

	}
}

File: refCsharp/MoveSet.cs
using System.Collections.Generic;

public class MoveSet
{
    // PROPERTIES
    private HashSet<Move> moves;

    // GETTERS
    public HashSet<Move> Moves { get { return moves; } }

    // CONSTRUCTORS
    public MoveSet()
    { moves = new HashSet<Move>(); }

    // METHODS
    public bool Add(Move move)
    { return moves.Add(move); }
    public bool IsEmpty()
    { return moves.Count == 0; }
    public Move GetMoveForPosition(IntVector position)
    {
        foreach (Move move in moves)
        {
            if (move == position) return move;
        }
        return null;
    }
    public void Purge(BaseBoard board, BasePiece piece)
    {
        // we must use an iterator because we are removing elements
        Purger predicate = new Purger(board, piece);
        moves.RemoveWhere(predicate.Purge);
    }

    // OVERRIDES
    public override string ToString()
    { return "MoveSet: " + moves.Count; }

    // OPERATORS
    public static bool operator +(MoveSet moves, Move move)
    { return moves.Add(move); }
    public static bool operator ~(MoveSet moves)
    { return moves.IsEmpty(); }

    // HELPERS
}

// class to remove elements while iterating
public struct Purger
{
    private BaseBoard board;
    private BasePiece piece;

    public Purger(BaseBoard board, BasePiece piece)
    {
        this.board = board;
        this.piece = piece;
    }

    public System.Predicate<Move> Purge { get { return _Test; } }

    private bool _Test(Move move)
    {
        BaseBoard tempBoard = (BaseBoard)board.Clone();
        BasePiece tempPiece = (BasePiece)piece.Clone();
        move.Apply(tempBoard, tempPiece);
        if (tempBoard.SingleKing && tempBoard.ThreatenKing(tempPiece.Team))
            return true;
        else
        {
            uint count;
            HashSet<BasePiece> kings = 
                tempBoard.ThreatenKings(tempPiece.Team, out count);
            if (kings.Count == count) return true;
        }
        return false;
    }
}